BE Report新規作成処理

tagahei opened the pull request:
## 対応内容・対応背景・妥協点
<!-- PRのざっくりした概要説明 -->
レポートの内容をDBに格納できる大枠ができた
その処理の単体テストを実装した

## やったこと
<!-- このPR内でやったことを箇条書きでまとめていく -->
- レポートの内容を受け取ったとき，それをDBに格納する
- 受け取ったものによってはエラーを返す
- その他細かい問題（テーブルデータの初期化等）の対応
- レポート格納処理のそれぞれの層での処理について，単体テストを実装

## やってないこと
<!-- このPR内でやっていないことを書く -->
- ユーザ関連は仮の状態

## テストの変更点
- [x] ある
- [ ] ない

## テストの方法
<!-- テスト項目、テスト方法を書く -->

next-appコンテナ内で以下のコマンドを実行
```sh
yarn test
```


## 実行例
成功している例
```sh
% curl -X POST -H "Content-Type: application/json" -d '{    "title": "サンプルタイトル",    "seminarDate": "2022-01-01",    "reportNumber": 1,    "pageNumber": 1,    "totalPages": 1,    "content": "レポートの内容",    "isSubmitted": false}' http://localhost:3000/api/v1/abc/report
Report has been created.%     
```

学生番号長いやつとかはこうなるよ
```sh
% curl -X POST -H "Content-Type: application/json" -d '{    "title": "サンプルタイトル",    "seminarDate": "2022-01-01",    "reportNumber": 1,    "pageNumber": 1,    "totalPages": 1,    "content": "レポートの内容",    "isSubmitted": false}' http://localhost:3000/api/v1/abcabcabc/report
Student ID must be less than 8 characters.%   
```


## レビューしてほしい点
- 

## 補足

[Prisma](https://www.prisma.io/)による型生成のため，初回はnext-appコンテナ内で以下のコマンドを実行

DBの内容からprismaのSchemaの作成
```sh
npx prisma db pull --schema=./src/lib/prisma/schema.prisma
```
SchemaからPrismaの型生成
```sh
npx prisma generate --schema=./src/lib/prisma/schema.prisma
```

tagahei committed:
add: テーブルの作成とprisma実行
Userテーブルはまだ使わないが，外部キーのために一応作りました．

Co-authored-by: Morikawa Seiya <s256270@users.noreply.github.com>

tagahei committed:
add: モックデータからレポートの新規作成
本来はドメインオブジェクトを受け取るが，createReport()を実装前のため，DBに必要な情報をいれるinsertReportを実装した

Co-authored-by: Morikawa Seiya <s256270@users.noreply.github.com>

tagahei committed:
add: ドメイン層におけるReportの作成と，DB保存呼び出し
Reportに必要なgetメソッドの追加しました
DBへの入力でUserへの外部キーが難しそうだったので，排除して実装を行なっています

Co-authored-by: Morikawa Seiya <s256270@users.noreply.github.com>

tagahei committed:
add: リクエストデータなしでのレポート新規作成
Userオブジェクトは一時的にアプリケーション層で作成した

Co-authored-by: Morikawa Seiya <s256270@users.noreply.github.com>

tagahei committed:
add: prismaのコマンドのメモ追加

tagahei committed:
add: Report作成に必要なリクエストの受け取り
Reportは確実にUserに紐づくものであり，studentIdは必須である．そのためstudentIdは，Pathパラメータから取得します．

Co-authored-by: Morikawa Seiya <s256270@users.noreply.github.com>

tagahei committed:
add: ドメイン層でUserの作成
本来はstudentIdを使ってDBからUser情報を取得するが，User関連の実装がまだなため，とりあえず仮のUser作成してます

Co-authored-by: Morikawa Seiya <s256270@users.noreply.github.com>

tagahei committed:
add: リクエストデータを受け取ってDBに格納
エンドポイントからDB格納までのデータのやり取りの繋げました．
Reportのコンストラクタとか型宣言して扱いやすいようにしています．

Co-authored-by: Morikawa Seiya <s256270@users.noreply.github.com>

tagahei committed:
update: DBに格納するときに成功/失敗を分かりやすくした
ユーザがないとき，学生番号が長すぎるときはクライアント側のミスとして400
それ以外のエラーはとりあえず500
特に問題ないときは200

Co-authored-by: Taga <banamon@users.noreply.github.com>

moririnrin committed:
fix: テーブルデータの初期化
DBのコンテナを起動したときに最初に読み込まれるSQLをマウントする
これによって，開発者全員が最初にDBの設定をできる

moririnrin committed:
fix: コンテナ起動時のエラー解消
```
failed to solve: Canceled: context canceled
```
上記のエラーが出たのでそれを解決するため

tkming0916 commented on next-app/src/lib/prisma/README.md:
```diff
@@ -0,0 +1,11 @@
+prismaのコマンド
```
このコマンドをいつ使うのか？
なんのためのコマンドなのかわからない...
もう少し丁寧に書いてください〜

tagahei replied:
aa4ff95
かいたちょ

tkming0916 replied:
見た👀

tkming0916 commented:
@banamon @s256270 
.envのサンプルがないんですけど〜〜〜ぉ

moririnrin committed:
add: .env.localのサンプルを追加
無いと指摘されたので追加した

moririnrin commented:
@tkming0916 
[676c3e7](https://github.com/YoslabEngineers/seminar-report-editor/pull/22/commits/676c3e779ccd13de1b3c40096046a68018f69d00) で.envのサンプルを追加しました

tkming0916 reviewed (changes requested):
割と良さげに見えました．
リードーミーだけもうちょいわかりやすく書いたほうがよきかな！

tagahei committed:
update: Prismaのコマンドについての説明を追加

Co-authored-by: Morikawa Seiya <s256270@users.noreply.github.com>

tagahei committed:
add: 成功した時のレスポンスを返す

元々，ドメインモデルをDBに格納していたが，ドメインモデルにもIDが欲しいため，DBに入れた後でドメインモデルを作成することにしたよ．

また，prismaのDBの型をドメインに変換は，新しく追加したインフラ層のFactoryで行っているよ．

Co-authored-by: Morikawa Seiya <s256270@users.noreply.github.com>

tagahei committed:
update: ReportにIDを追加

 Reportを一意に定めるために，DBで格納した時のIDを取得してドメインモデルが持つようにしたよぉ

tagahei committed:
ref: 成功レスポンスをtrue/falseで判定

文字列比較で成功かどうか判定していたが，createReportの返り値のぬる判定で判断するよぉ

後，201番が新しくリソースを追加した時のHTTP Statusらしい

tagahei committed:
update: createUserからgetUserに変更

Reportの新規作成では，Userを作成するのでなく，DBの中のUser情報からUserドメインモデルを生成する感じだよね

tagahei committed:
refactor: FactoryでReportのドメインモデルの生成

tagahei committed:
update: エラーハンドリング

User,Reportの各処理で発生するエラーを投げたよぉ

tagahei committed:
add: ユーザが見つからなかった時の例外処理

Co-authored-by: Morikawa Seiya <s256270@users.noreply.github.com>

sueCalystegia00 reviewed (approved):
LGTM

tagahei committed:
add: createReportのテストの追加

jestのエイリアスがまだうまく設定できていないため，相対パスでimportしましょう
また，jestもVSCodeで読み込めていないのでエラー出てますが，実行できます

tagahei committed:
add: registerReportのテストの追加

tagahei committed:
refactor: mock対象のpathを絶対pathで記述

tagahei committed:
refactor: Give-When-Then構文がわかりやすいようにコメント追加

tagahei committed:
add: prisma用のJestの拡張パッケージ

```
yarn add jest-mock-extended@2.0.4 --dev
```

tagahei committed:
refactor: prismaのインスタンスをlibで生成

tagahei committed:
fix: Userドメインモデルにidの追加

ReportのUser情報が外部キーとしてusersのidを見ているから，ドメインモデルが持つ方がいいよね

tagahei committed:
add: prisma clientのシングルトン化

tagahei committed:
fix: prismaの型からFactoryでドメインモデルの変化

元々anyだったのを明示的に変換しました

Co-authored-by: Sue <sueCalystegia00@users.noreply.github.com>

sueCalystegia00 committed:
fix: table名の先頭を大文字に変更

prismaのモデルとして管理されるため大文字の方が型として扱いやすい

sueCalystegia00 committed:
refactor: Prismaの型利用を明示

sueCalystegia00 committed:
remove: 不要なコードの削除

動作確認用に作ったもの

sueCalystegia00 committed:
upgrade: jest型パッケージの追加と拡張パッケージのアップグレード

sueCalystegia00 committed:
wip: reports.create()の単体テスト

型を整理したところmockResolvedValueがないことがわかる。
つまりsingleton.tsで定義したprismaMockが想定していたjest.mockとしての動きをしていない。
prismaの公式ドキュメントに従いjestの拡張パッケージを利用して実装していたが上手くいかないとなると、根本的にモックの建て方を返るべきかも

tagahei committed:
remove: prismaのmock化を諦めた

prismaの型変換がむずいねぇ

tagahei committed:
update: DB直接殴ってテスト

prismaをフロントではなくバックで動かすためにjest.configの修正したよ
とりあえず，今のDB殴るようにした．

tagahei committed:
update: Userテーブル初期化時のサンプルデータ挿入

tagahei committed:
add: test終わりにReportsテーブルのクリーンアップ

テストでReportIdを比較したいが，IDはオートインクリメントであるため，テスト毎にテーブルのクリーンアップを行う．
jest.setup.tsに書き出して全てのテスト終わりに実行されます

moririnrin committed:
fix: プリズマの型の更新

テーブル名を変更し，プリズマのスキーマまで変更しているが，型生成ができていなかったため，生成した．
それに伴った，型名などの変更
以下の2つでプリズマの型生成
```sh
npx prisma db pull --schema=./src/lib/prisma/schema.prisma
```
```sh
npx prisma generate --schema=./src/lib/prisma/schema.prisma
```

moririnrin committed:
add: テスト時にUsersテーブルに仮の値を入れて，最後に初期化する

動くけど，3回のテストごとに仮データの挿入と初期化が実行される

Co-authored-by: Taga <banamon@users.noreply.github.com>

moririnrin committed:
add: テスト用のデータベースを作った

なぜかsequel aceからは入れないが，dockerからは観れるので一旦このまま進んでみる

Co-authored-by: Taga <banamon@users.noreply.github.com>

moririnrin committed:
fix: コンテナ側のポート番号を変更

portsの左右はローカルとコンテナに対応している．
MySQLのイメージは，デフォルトで3306を開放するので，ローカルは3307のまま，コンテナ側を3306に設定するのが正しい

moririnrin committed:
add: テスト実行時に環境変数を切り替える

テスト実行時に参照するデータベースをテスト用のものに切り替えたいため

moririnrin committed:
update: テーブルの仮データ挿入とクリーンアップを必要な時のみ行うようにした

全てのテストのたびに動作していたので，データベース操作に関係のあるテストの時のみに動作するようにした

moririnrin committed:
refactor: データベースの初期化時に仮のデータを入れないようにした

動作確認はテストで行うようにしたので，サンプルを入れておく必要がなくなったため

tagahei committed:
add: テストに対応するサンプルenvファイル

Co-authored-by: Morikawa Seiya <s256270@users.noreply.github.com>

tagahei committed:
refactor: @でimport

Co-authored-by: Morikawa Seiya <s256270@users.noreply.github.com>

tagahei committed:
add: パッケージの追加

APIテスト用のパッケージ
```sh
yarn add --dev next-test-api-route-handler
```

tagahei committed:
add: Report作成APIの正常系テスト

Co-authored-by: Morikawa Seiya <s256270@users.noreply.github.com>

tagahei committed:
update: Table Driven Testsで記述

registerNewReportの返り値のみ違うテストのため簡単に書いた

Co-authored-by: Morikawa Seiya <s256270@users.noreply.github.com>

tagahei committed:
refactor: サーバーエラメッセ時の変更

Co-authored-by: Morikawa Seiya <s256270@users.noreply.github.com>

tagahei committed:
add: Report作成APIの異常系テスト

Co-authored-by: Morikawa Seiya <s256270@users.noreply.github.com>

tagahei committed:
refactor: Table Driven Testで記述

Co-authored-by: Morikawa Seiya <s256270@users.noreply.github.com>

tagahei committed:
refactor: 不要なコンソール削除

tagahei committed:
fix: コンフリ解消ミスの対処

sueCalystegia00 reviewed (approved):
モブプロで見てたのもあるので、大枠問題ないかと

tkming0916 commented:
ざっと見ました！
こっちの環境で動かしてみようとしたけど，`npx prisma db pull --schema=./src/lib/prisma/schema.prisma` でenv内のURLがないって言われちゃった🥺

tagahei commented:
> ざっと見ました！ こっちの環境で動かしてみようとしたけど，`npx prisma db pull --schema=./src/lib/prisma/schema.prisma` でenv内のURLがないって言われちゃった🥺

確認あざす！
.env.local.sample参考に編集してもろたらいけるかと！ディスコに投げとくわ

tagahei committed:
fix: ドメインモデルの変更をフロントに反映

tagahei committed:
refactor: ディレクトリ革命

tagahei committed:
refactor: appとapplicationの統合

sueCalystegia00 committed:
Merge pull request #38 from YoslabEngineers/refactor/architecture

Refactor/architecture
